apps/emqx_resource/src/emqx_resource_pool.erl

```diff
@@ -66,27 +67,40 @@ stop(Name) ->
     end.
 
 health_check_workers(PoolName, CheckFunc) ->
-    health_check_workers(PoolName, CheckFunc, ?HEALTH_CHECK_TIMEOUT).
+    health_check_workers(PoolName, CheckFunc, ?HEALTH_CHECK_TIMEOUT, _Opts = #{}).
 
 health_check_workers(PoolName, CheckFunc, Timeout) ->
+    health_check_workers(PoolName, CheckFunc, Timeout, _Opts = #{}).
+
+health_check_workers(PoolName, CheckFunc, Timeout, Opts) ->
+    ReturnValues = maps:get(return_values, Opts, false),
     Workers = [Worker || {_WorkerName, Worker} <- ecpool:workers(PoolName)],
     DoPerWorker =
         fun(Worker) ->
             case ecpool_worker:client(Worker) of
                 {ok, Conn} ->
                     erlang:is_process_alive(Conn) andalso
                         ecpool_worker:exec(Worker, CheckFunc, Timeout);
-                _ ->
-                    false
+                Error ->
+                    Error
             end
         end,
-    try emqx_utils:pmap(DoPerWorker, Workers, Timeout) of
-        [_ | _] = Status ->
-            lists:all(fun(St) -> St =:= true end, Status);
-        [] ->
-            false
-    catch
-        exit:timeout ->
+    Results =
+        try
+            emqx_utils:pmap(DoPerWorker, Workers, Timeout)
+        catch
+            exit:timeout ->
+                {error, timeout}
+        end,
+    case ReturnValues of
+        true when is_list(Results) ->
+            {ok, Results};
+        true ->
+            Results;
```

Comment:
is there some reason to differentiate the return based on `Results` when `ReturnValues` is true?

Reply:
yes, it may not be a successful list: https://github.com/emqx/emqx/pull/11107/files#diff-d9d2be3536c8645f4b9a4eef5702ee6b820aa98047b6d1059b92d3498f8e99f3R93

Reply:
but why do not return `{ok, Results}` for both cases?

Reply:
it's not equivalent.  but I'll refactor it to avoid checking on the type.